README.md
```
         __
        /  \
       / ..|\
      (_\  |_)
      /  \@'    Alberto!
     /     \         0.5
_  _/  `   |
\\/  \  | _\
 \   /_ || \\_
  \____)|_) \_)
```

Alberto is an implementation of
[Erlang External Term Format](http://erlang.org/doc/apps/erts/erl_ext_dist.html),
a protocol, used by Erlang nodes to communicate with so called *ports*. See
Erlang [documentation](http://www.erlang.org/doc/tutorial/c_port.html) for
details.

Installation
------------

The simplest way to get `alberto` is to use [`OPAM`](http://opam.ocamlpro.com):

```bash
opam install alberto
```

Example
-------

On Erlang part we have a simple echo server, which sends `<<"Hello world!">>`
to the port and echoes whatever comes back to stdout:

```erlang
#!/usr/bin/env escript
%%! -noshell -noinput
-mode(compile).
-define(ME, filename:basename(escript:script_name())).
-define(PRINT(STR, PARAMS), io:format("~s: " ++ STR ++ "~n", [?ME | PARAMS])).


main([Bin]) ->
    Port = open_port({spawn, Bin},
                     [binary, {packet, 4}]),
    port_command(Port, term_to_binary("Hello world!")),
    receive
        {_, {data, Data}} ->
            ?PRINT("~p", [binary_to_term(Data)]),
            main([Bin])
    end;
main(_) ->
    io:format("~s", ["usage: echo.erl path/to/port"]).
```

OCaml part is even more consice:

```ocaml
Alberto.interact (fun x -> x)
```

Okay, let's see it in action (sources are available in `examples/` directory):

```bash
$ make
$ examples/echo.erl ./port_simple.native
echo.erl: "Hello world!"
echo.erl: "Hello world!"
echo.erl: "Hello world!"
echo.erl: "Hello world!"
```

**Note**: to build examples in the `examples/` directory configure
`alberto` with `--enable-examples` flag.


CHANGES
alberto Changelog
=================

Here you can see the full list of changes between each alberto release.

Version 0.5
-----------

Maintenance release, released on November 21st, 2017

- Removed dependency on Camlp4
- Switched from ocamlbuild and OASIS to jbuilder

Version 0.4
-----------

Bugfix release, released on March 3rd, 2015

- Fixed an off-by-one error introduced after migration to 'safe_string'
- Disabled warnings-as-errors

Version 0.3
-----------

Bugfix release, relased on December 23rd, 2014

- Eliminated warnings produced by OCaml 4.02.

Version 0.2
-----------

Bugfix release, released on May 13th, 2014

- Switched to 'ocplib-endian' for serialization of integral types.


Version 0.1
-----------

Initial release, released on October 25th, 2011


tests.ml
(** Tests for Alberto module. *)

open Kaputt.Abbreviations


module String = struct
  include String

  let resize s n =
    if length s > n
    then sub s 0 n
    else
      let buf = Bytes.make n '\000' in
      blit s 0 buf 0 (length s);
      Bytes.to_string buf
end


module ErlGen = struct
  module G = Gen

  (* Helpers. *)
  let pos_int8  = G.make_int 0 (1 lsl 8  - 1)
  let pos_int16 = G.make_int 0 (1 lsl 16 - 1)

  (* Primitive types. *)
  let small_int = G.map1 (fun x -> `Int x) Alberto.to_string pos_int8
  let integer = G.map1 (fun x -> `Int x) Alberto.to_string
    (G.map1 Int32.to_int string_of_int G.int32)
  let atom = G.map1 (fun s -> `Atom s) Alberto.to_string (G.word pos_int8)
  let float = G.map1 (fun f -> `Float f) Alberto.to_string G.float
  let reference = G.map1 (fun r -> `Reference r) Alberto.to_string
    G.(zip3
       (word pos_int8)
       (word (make_int 0 16))
       pos_int8)
  let port = G.map1 (fun p -> `Port p) Alberto.to_string
    G.(zip3
       (word pos_int8)
       (map1 Int32.to_int string_of_int pos_int32)
       pos_int8)
  let pid = G.map1 (fun p -> `PID p) Alberto.to_string
    G.(zip4
       (word pos_int8)
       (word (make_int 1 4))
       (map1 Int32.to_int string_of_int int32)
       pos_int8)
  let nil = G.map1 (fun _ -> `List []) Alberto.to_string G.unit
  let small_string = G.map1 (fun s -> `String s) Alberto.to_string
    (G.word pos_int16)
  let large_string = G.map1 (fun s -> `String s) Alberto.to_string
    G.(word (make_int (1 lsl 16) (1 lsl 20)))
  let binary = G.map1 (fun b -> `Binary b) Alberto.to_string
    (G.buffer (G.word pos_int16))
  let small_big = G.map1 (fun n -> `Bignum n) Alberto.to_string
    (KaputtNums.Generator.big_int pos_int8)
  let large_big = G.map1 (fun n -> `Bignum n) Alberto.to_string
    (KaputtNums.Generator.big_int pos_int16)
  let new_reference = G.map1
    (fun (n, c) -> `NewReference
        (n, c, String.(make ((length n) * 4) '\000')))
    Alberto.to_string
    G.(zip2 (word pos_int8) pos_int8)
  let bit_binary = G.map1 (fun b -> `BitBinary b) Alberto.to_string
    G.(zip2
         (G.buffer (word pos_int8))
         pos_int8)

  (* Compound types. *)
  let term = G.choose_list [small_int; integer; atom; float; port; nil]

  let small_tuple = G.map1 (fun l -> `Tuple l) Alberto.to_string
    (G.list pos_int8 term)
  let large_tuple = G.map1 (fun l -> `Tuple l) Alberto.to_string
    (G.list pos_int16 term)
  let list = G.map1 (fun l -> `List l) Alberto.to_string (G.list pos_int16 term)
end


module Spec = struct
  include Spec

  let id (x, y) = x = y
end


let parse_unparse term = Alberto.decode_exn (Alberto.encode_exn term);;


Test.add_random_test
  ~title:"SMALL_INTEGER_EXT"
  ~nb_runs:500
  ErlGen.small_int
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"INTEGER_EXT"
  ~nb_runs:500
  ErlGen.integer
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"FLOAT_EXT"
  ~nb_runs:500
  ErlGen.float
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"ATOM_EXT"
  ~nb_runs:500
  ErlGen.atom
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"REFERENCE_EXT"
  ErlGen.reference
  parse_unparse
  [Spec.always => function
    | (`Reference (node, id, creation), r) ->
      r = (`Reference (node, (String.resize id 4), creation))
    | (_, _) -> false]
;;


Test.add_random_test
  ~title:"PORT_EXT"
  ErlGen.port
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"PID_EXT"
  ErlGen.pid
  parse_unparse
  [Spec.always => function
    | (`PID (node, id, serial, creation), p) ->
      p = (`PID (node, (String.resize id 4), serial, creation))
    | (_, _) -> false]
;;


Test.add_random_test
  ~title:"SMALL_TUPLE_EXT"
  ErlGen.small_tuple
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"LARGE_TUPLE_EXT"
  ~nb_runs:10
  ErlGen.large_tuple
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"NIL_EXT"
  ErlGen.nil
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"STRING_EXT (small)"
  ~nb_runs:500
  ErlGen.small_string
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"STRING_EXT (large)"
  ~nb_runs:10
  ErlGen.large_string
  parse_unparse
  [Spec.always => function
     | (`String _, `List _) -> true
     | _ -> false]
;;


Test.add_random_test
  ~title:"LIST_EXT"
  ~nb_runs:10
  ErlGen.list
  parse_unparse
  [Spec.always => Spec.id]
;;


Test.add_random_test
  ~title:"BINARY_EXT"
  ErlGen.binary
  parse_unparse
  [Spec.always => function
    | (`Binary b, `Binary b') -> (Buffer.contents b) = (Buffer.contents b')
    | _ -> false]
;;


Test.add_random_test
  ~title:"SMALL_BIG_EXT"
  ErlGen.small_big
  parse_unparse
  [Spec.always => function
    | (`Bignum n, `Bignum n') -> Big_int.eq_big_int n n'
    | _ -> false]
;;


Test.add_random_test
  ~title:"LARGE_BIG_EXT"
  ~nb_runs:10
  ErlGen.large_big
  parse_unparse
  [Spec.always => function
    | (`Bignum n, `Bignum n') -> Big_int.eq_big_int n n'
    | _ -> false]
;;


Test.add_random_test
  ~title:"NEW_REFERENCE_EXT"
  ErlGen.new_reference
  parse_unparse
  [Spec.always => function
    | (`NewReference (node, creation, id), r) ->
      let len = (String.length id) / 4 in
      r = (`NewReference (node, creation, (String.resize id (len * 4))))
    | (_, _) -> false]
;;


Test.add_random_test
  ~title:"BIT_BINARY_EXT"
  ErlGen.bit_binary
  parse_unparse
  [Spec.always => function
    | (`BitBinary (b, 0), `Binary b') ->
      (Buffer.contents b) = (Buffer.contents b')
    | (`BitBinary (b, bits), `BitBinary (b', bits')) ->
      (Buffer.contents b) = (Buffer.contents b') && bits = bits'
    | _ -> false]
;;


let () =
  Test.launch_tests ()
;;


examples/port_lwt.ml
(** Erlang ports, the Lwt way! *)

module Alberto_lwt : sig
  val read_term : Lwt_io.input_channel -> Alberto.t Lwt.t
  (** [read_term ic] fetches an Erlang [!term] from the port channel [ic]. *)

  val write_term : Lwt_io.output_channel -> Alberto.t -> unit Lwt.t
  (** [write_term oc term] serializes a given Erlang [!term] to the port
      channel [oc]. *)

  val interact : (Alberto.t -> Alberto.t) -> unit Lwt.t
  (** [interact f] main port loop, recieves Erlang terms from stdin,
      executes a given key function [f] and sends back the results. *)
end = struct
  open Lwt

  let rec read_term ic =
    Lwt_io.BE.read_int ic >>= fun len ->
    (* The current version of lwt doesn't use -safe-string. *)
    let buf = Bytes.create len in
    Lwt_io.read_into_exactly ic buf 0 len >>= fun _ ->
    return @@ Alberto.decode_exn (Bytes.to_string buf)

  and write_term oc term =
    let buf = Alberto.encode_exn term in
    Lwt_io.BE.write_int oc (String.length buf) >>= fun () ->
    Lwt_io.write oc buf >>= fun () -> Lwt_io.flush oc

  and interact f =
    let rec transform state = function
      | `Atom "stop" -> exit 0
      | `Atom "ping" -> state, `Atom "pong"
      | term ->
        let term = try f term with exn ->
          `Tuple [`Atom "error";
                  `Tuple [term; `String (Printexc.to_string exn)]]
        in (state, term)
    and loop state =
      read_term Lwt_io.stdin >>= fun term ->
      let (state, term) = transform state term in
      write_term Lwt_io.stdout term >>= fun () -> loop state
    in Lwt.catch (fun () -> loop 0) (fun exn ->
        match exn with
          | End_of_file -> exit 1
          | _ -> raise exn)
end


let () = Lwt_main.run (Alberto_lwt.interact (fun x -> x))


src/alberto.ml
open Big_int

type t = [ `Int of int
         | `Float of float
         | `Atom of string
         | `Reference of (string * string * int)
         | `Port of (string * int * int)
         | `PID of (string * string * int * int)
         | `Tuple of t list
         | `List of t list
         | `String of string
         | `Binary of Buffer.t
         | `Bignum of big_int
         | `NewReference of (string * int * string)
         | `BitBinary of (Buffer.t * int)
         ]

(* +-------+
   | Utils |
   +-------+ *)

module BE = EndianString.BigEndian

let (<| ) f x = f x

let failwithf fmt = Printf.kprintf failwith fmt


(** Shortcuts for some of the binary functions. *)
let pack_byte n =
  let buf = Bytes.create 1 in begin
    BE.set_int8 buf 0 n;
    buf
  end
and pack_word n =
  let buf = Bytes.create 2 in begin
    BE.set_int16 buf 0 n;
    buf
  end
and pack_int32 n =
  let buf = Bytes.create 4 in begin
    BE.set_int32 buf 0 (Int32.of_int n);
    buf
  end


(** [big_int_digits n] returns a list of digits of a given number [n]. *)
let big_int_digits n =
  let offset = big_int_of_int 256 in
  let rec aux n l =
    if eq_big_int n zero_big_int then
      l
    else
      let q, m = quomod_big_int n offset in
      aux q (int_of_big_int m :: l)
  in aux n []


module Stream = struct
  include Stream

  (** [take n st] returns a *sub-stream* of stream [st], which has [n]
      or fewer characters. *)
  let rec take n st =
    if n > 0 then
      slazy (fun _ -> take (n - 1) st) |> icons (next st)
    else
      sempty

  (** [to_string ?length st] returns string representation of a given
      stream [st], using a buffer of a given [length] for conversion. *)
  let to_string ?(length=64) st =
    let buf = Buffer.create length in begin
      iter (fun c -> Buffer.add_char buf c) st;
      Buffer.contents buf
    end

  (** [take_string len st] extracts and returns a string of a given [length]
      from stream [st]. *)
  let take_string length st = take length st |> to_string ~length

  (** [take_byte st] extracts and returns a 1 byte unsigned integer
      from stream [st]. *)
  let take_byte st = BE.get_uint8 (take_string 1 st) 0

  (** [take_word st] extracts and returns a 2 byte unsigned integer
      from stream [st]. *)
  let take_word st = BE.get_uint16 (take_string 2 st) 0

  (** [take_int32 st] extracts and returns a 4 byte signed integer from
      stream [st]. *)
  let take_int32 st = BE.get_int32 (take_string 4 st) 0 |> Int32.to_int

  (** [take_big len st] extracts and returns a [!Bignum] of length
      [len] from stream [st]. *)
  let take_big len st =
    let offset = 256 in
    let rec aux idx n = match idx with
      | 0 -> n
      | _ ->
        if idx < 0 then
          failwithf "Invalid length: %i" idx
        else begin
          let digit = take_byte st in
          aux (idx - 1)
            (add_int_big_int digit (mult_int_big_int offset n))
        end
    in aux len (big_int_of_int 0)

  (** [take_list len f st] extracts and returns a [!List] of items
      fetched by [f] from stream [st]. *)
  let take_list len f st =
    let rec aux idx l = match idx with
      | 0 -> List.rev l
      | _ ->
        if idx < 0 then
          failwithf "Invalid length: %i" idx
        else (f st) :: l |> aux (idx - 1)
    in aux len []
end


module String = struct
  include String

  let resize s n =
    let buf = Bytes.make n '\000' in

    if length s < n then
      blit s 0 buf 0 (length s)
    else
      blit s 0 buf 0 n;

    Bytes.to_string buf
end


(* +--------+
   | Parser |
   +--------+ *)

module S = Stream

let rec parse s = match S.next s with
  (* SMALL_INTEGER_EXT: Unsigned 8 bit integer. *)
  | '\097' -> `Int (S.take_byte s)

  (* INTEGER_EXT: Signed 32 bit integer in big-endian format, i.e.
     MSB first. *)
  | '\098' -> `Int (S.take_int32 s)

  (* FLOAT_EXT: A float is stored in string format. the format used
     in sprintf to format the float is "%.20e". *)
  | '\099' ->
     let blob = S.take_string 31 s in
     let zeros = String.index blob '\000' in
     `Float (float_of_string <| String.sub blob 0 zeros)

  (* ATOM_EXT *)
  | '\100' ->
     let len = S.take_word s in
     if len < 256
     then `Atom (S.take_string len s)
     else failwithf "ATOM_EXT length exceeded %i > 256" len

  (* REFERENCE_EXT *)
  | '\101' ->
     let node = parse s in
     let id = S.take_string 4 s in
     let creation = S.take_byte s in
     let node = match node with
       | `Atom s -> s
       | #t -> failwithf "Unexpected REFERENCE_EXT format"
     in `Reference (node, id, creation)

  (* PORT_EXT *)
  | '\102' ->
     let node = parse s in
     let id = S.take_int32 s in
     let creation = S.take_byte s in
     let node = match node with
       | `Atom s -> s
       | #t -> failwithf "Unexpected PORT_EXT format"
     in `Port (node, id, creation)

  (* PID_EXT *)
  | '\103' ->
    let node = parse s in
    let id = S.take_string 4 s in
    let serial = S.take_int32 s in
    let creation = S.take_byte s in
    let node = match node with
      | `Atom s -> s
      | #t -> failwithf "Unexpected PID_EXT format"
    in `PID (node, id, serial, creation)

  (* SMALL_TUPLE_EXT *)
  | '\104' ->
     let arity = S.take_byte s in
     let elements = S.take_list arity parse s in
    `Tuple elements
  (* LARGE_TUPLE_EXT *)
  | '\105' ->
    let arity = S.take_int32 s in
    let elements = S.take_list arity parse s in
    `Tuple elements

  (* NIL_EXT *)
  | '\106' -> `List []
  (* LIST_EXT *)
  | '\108' ->
     let len = S.take_int32 s in
     let elements = S.take_list len parse s in
     begin match S.next s with
     (* NIL. *)
     | '\106' -> `List elements
     | _ -> failwith "Unexpected LIST_EXT format"
     end

  (* STRING_EXT *)
  | '\107' ->
     let len = S.take_word s in
     let blob = S.take_string len s in
     `String blob

  (* BINARY_EXT *)
  | '\109' ->
    let len = S.take_int32 s in
    let blob = S.take_string len s in
    let buf = Buffer.create len in begin
      Buffer.add_string buf blob;
      `Binary buf
    end

  (* SMALL_BIG_EXT *)
  | '\110' ->
    let len = S.take_byte s in
    let sign = S.take_byte s in
    let n = S.take_big len s in
    `Bignum (if sign > 0 then minus_big_int n else n)
  (* LARGE_BIG_EXT *)
  | '\111' ->
     let len = S.take_int32 s in
     let sign = S.take_byte s in
     let n = S.take_big len s in
    `Bignum (if sign > 0 then minus_big_int n else n)

  (* NEW_REFERENCE_EXT *)
  | '\114' ->
    let len = S.take_word s in
    let node = parse s in
    let creation = S.take_byte s in
    let id = S.take_string (len * 4) s in
    let node = match node with
      | `Atom s -> s
      | #t -> failwithf "Unexpected NEW_REFERENCE_EXT format"
    in `NewReference (node, creation, id)

  (* FUN_EXT *)
  (* NEW_FUN_EXT *)
  (* EXPORT_EXT *)

  (* BIT_BINARY_EXT *)
  | '\077' ->
    let len = S.take_int32 s in
    let bits = S.take_byte s in
    let blob = S.take_string len s in
    let buf = Buffer.create len in begin
      Buffer.add_string buf blob;
      `BitBinary (buf, bits)
    end

  (* Unknown tag? *)
  | tag -> invalid_arg (String.make 1 tag)


(* +------------+
   | Serializer |
   +------------+ *)

let rec serialize buf term =
  let add f x =
    let res = f x in
    (* We have to use explicit loop, because of the Buffer API change
       in OCaml>=4.02. *)
    for i = 0 to Bytes.length res - 1 do
      Buffer.add_char buf (Bytes.get res i)
    done
  in

  let open Buffer in match term with
    | `Int n when n >= 0 && n < 256 ->
      begin add pack_byte 97; add pack_byte n end
    | `Int n ->
      begin add pack_byte 98; add pack_int32 n end
    | `Float f ->
      let s = Printf.sprintf "%.20e" f in begin
        add pack_byte 99;
        add_string buf s;
        add_string buf <| String.make (31 - String.length s) '\000'
      end
    | `Atom s ->
      let length = String.length s in
      if length < 256 then (
        add pack_byte 100;
        add pack_word length;
        add_string buf s
      ) else
        failwithf "ATOM_EXT length exceeded %i > 255" length
    | `Reference (node, id, creation) ->
      begin
        add pack_byte 101;
        serialize buf (`Atom node); (* reuse the buf we already have. *)
        Buffer.add_string buf (String.resize id 4);
        add pack_byte creation
      end
    | `Port (node, id, creation) ->
      begin
        add pack_byte 102;
        serialize buf (`Atom node);
        add pack_int32 id;
        add pack_byte creation
      end
    | `PID (node, id, serial, creation) ->
        add pack_byte 103;
        serialize buf (`Atom node);
        Buffer.add_string buf (String.resize id 4);
        add pack_int32 serial;
        add pack_byte creation
    | `Tuple l ->
      let length = List.length l in
      if length < 256 then (
        add pack_byte 104;
        add pack_byte length;
      ) else (
        add pack_byte 105;
        add pack_int32 length;
      );
      List.iter (serialize buf) l
    | `List [] -> add pack_byte 106
    | `List l ->
      begin
        add pack_byte 108;
        add pack_int32 (List.length l);
        List.iter (serialize buf) l;
        add pack_byte 106
      end
    | `String s ->
      let length = String.length s in
      if length < 65536 then (
        add pack_byte 107;
        add pack_word length;
        add_string buf s
      ) else (
          (* Strings longer than 65535 bytes are encoded as LIST_EXT.
             No explicit convertion to [List] is done, because it really
             slows things down for large strings. *)
        add pack_byte 108;
        add pack_int32 (String.length s);
        String.iter (fun c -> add_char buf '\097'; add_char buf c) s;
        add pack_byte 106
      )
    | `Binary b ->
      begin
        add pack_byte 109;
        add pack_int32 (Buffer.length b);
        add_buffer buf b
      end
    | `Bignum n ->
      let sign = match sign_big_int n with -1 -> 1 | _ -> 0 in
      let digits = big_int_digits (abs_big_int n) in
      let length = List.length digits in
      if length < 256 then (
        add pack_byte 110;
        add pack_byte length;
      ) else (
        add pack_byte 111;
        add pack_int32 length;
      );
      add pack_byte sign;
      List.iter (fun d -> add pack_byte d) digits
    | `NewReference (node, creation, id) ->
      let len = (String.length id) / 4 in
      add pack_byte 114;
      add pack_word len;
      serialize buf (`Atom node);
      add pack_byte creation;
      Buffer.add_string buf (String.resize id (len * 4))
    | `BitBinary (b, 0) -> serialize buf (`Binary b)
    | `BitBinary (b, bits) ->
      begin
        add pack_byte 77;
        add pack_int32 (Buffer.length b);
        add pack_byte bits;
        add_buffer buf b
      end


(* +------------+
   | Public API |
   +------------+ *)

let decode_exn s =
  match s.[0] with
    | '\131' ->
      let stream = S.of_string s in
      let (_ : char) = S.next stream in
      parse stream
    | n   ->
      failwithf "Erlang binary doesn't start with 131: %i" (int_of_char n)

and encode_exn term =
  let buf = Buffer.create 256 in begin
    Buffer.add_char buf '\131';
    serialize buf term;
    Buffer.contents buf
  end


let decode s =
  try
    Some (decode_exn s)
  with _ -> None

and encode term =
  try
    Some (encode_exn term)
  with _ -> None


(* +------------------+
 * | Pretty printers. |
 * +------------------+ *)

let rec to_string = function
  | `Int x -> string_of_int x
  | `Float f -> string_of_float f
  | `Atom s | `String s -> s
  | `Reference (node, id, creation) | `NewReference (node, creation, id) ->
    Printf.sprintf "#Ref<[%s]:%s.%i>" node id creation
  | `Port (node, id, creation) ->
    Printf.sprintf "#Port<[%s]:%i.%i>" node id creation
  | `PID (node, id, serial, creation) ->
    Printf.sprintf "#PID<[%s]:%s.%i.%i>" node id serial creation
  | `Tuple terms ->
    Printf.sprintf "{%s}" (String.concat "," (List.map to_string terms))
  | `List terms ->
    Printf.sprintf "[%s]" (String.concat "," (List.map to_string terms))
  | `Binary b -> Printf.sprintf "<<%s>>" (Buffer.contents b)
  | `Bignum n -> string_of_big_int n
  | `BitBinary (b, bits) -> Printf.sprintf "<<%s:%i>>" (Buffer.contents b) bits


(* +-----------------------------------------+
   | Here come Erlang ports! -- the easy way |
   +-----------------------------------------+ *)

let rec read_term ic =
  let len =
    try
      input_binary_int ic
    with End_of_file -> exit 0 in
  let buf = Bytes.create len in begin
    really_input ic buf 0 len;
    decode_exn (Bytes.to_string buf)
  end

and write_term oc term =
  let buf = encode_exn term in begin
    output_binary_int oc (String.length buf);
    output_string oc buf;
    flush oc
  end

and interact f =
  let rec transform state = function
    | `Atom "stop" -> exit 0
    | `Atom "ping" -> state, `Atom "pong"
    | term ->
      let term = try f term with exn ->
        `Tuple [`Atom "error";
                `Tuple [term; `String (Printexc.to_string exn)]]
      in (state, term)
  and loop state =
    let (state, term) = transform state <| read_term stdin in begin
      write_term stdout term;
      loop state
    end
  in begin
    set_binary_mode_in stdin true;
    set_binary_mode_out stdout true;

    try loop 0 with End_of_file -> exit 0;

  end
